mixer/adapter/signalfx/signalfx.go

```
@@ -0,0 +1,188 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/signalfx/golib/sfxclient"
+
+	"istio.io/istio/mixer/adapter/signalfx/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/metric"
+)
+
+type (
+	builder struct {
+		metricTypes map[string]*metric.Type
+		config      *config.Params
+	}
+
+	handler struct {
+		env                adapter.Env
+		ctx                context.Context
+		cancel             func()
+		scheduler          *sfxclient.Scheduler
+		registry           *registry
+		metricTypes        map[string]*metric.Type
+		ingestURL          string
+		accessToken        string
+		intervalSeconds    uint32
+		metricConfigByName map[string]*config.Params_MetricConfig
+		logger             adapter.Logger
+	}
+)
+
+// ensure types implement the requisite interfaces
+var _ metric.HandlerBuilder = &builder{}
+var _ metric.Handler = &handler{}
+
+///////////////// Configuration-time Methods ///////////////
+
+// adapter.HandlerBuilder#Build
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	confsByName := make(map[string]*config.Params_MetricConfig, len(b.config.Metrics))
+	for i := range b.config.Metrics {
+		confsByName[b.config.Metrics[i].Name] = b.config.Metrics[i]
+	}
+
+	env.Logger().Infof("Building SignalFx adapter")
+
+	h := &handler{
+		env:                env,
+		metricTypes:        b.metricTypes,
+		ingestURL:          b.config.IngestUrl,
+		accessToken:        b.config.AccessToken,
+		intervalSeconds:    b.config.IntervalSeconds,
+		metricConfigByName: confsByName,
+		logger:             env.Logger(),
+	}
+
+	return h, h.Init()
+}
+
+// adapter.HandlerBuilder#SetAdapterConfig
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.config = cfg.(*config.Params)
+
+	if b.config.IntervalSeconds == 0 {
+		b.config.IntervalSeconds = 10
+	}
+}
+
+// adapter.HandlerBuilder#Validate
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
```

You should also validate that:

- The URI is properly formatted.
- The supplied time interval value is in range
- All the metrics from b.metricTypes have corresponding config. Otherwise, Istio will be sending you these metrics at runtime and you won't have anything to do with them.

---

Oh and given the constraints in your valueToFloat function, you should check to ensure all of the metrics in b.metricTypes only hold dimensions of type you can deal with at runtime. We always want to fail early at config time rather than at runtime when possible.